### DISCUSSION TITLE
Hurrican - Port/compile for RG350/OpenDingux possible?

### DISCUSSION THREAD

**Discussion by agerhard0** [eduardofilo/RG350_adam_image#244] [Ports]
Hello together,

as the fan project Hurrican covering the original Turrican game mechanics is available for RG351-units, I would like to start the discussion, whether it is possible with the available code to create a port for RG350-units.

Project: https://www.winterworks.de/project/hurrican/

Thought I posted something like this already - but I am unable to find it again.

Found the following ports/sources - maybe someone with the skills can have a look into it:

Hurrican SDL Port / Raspberry Pi:
https://forums.raspberrypi.com/viewtopic.php?t=12932
https://www.phoronix.com/forums/forum/software/linux-gaming/29540-hurrican-sdl-port?69804-Hurrican-SDL-Port=
https://misapuntesde.com/post.php?id=242

Odroid Port:
https://github.com/meveric/hurrican

Pandora Port:
http://www.repo.openpandora.org/?page=detail&app=hurrican.pickle
https://sourceforge.net/projects/hurrican/
https://github.com/NebuPookins/hurrican

General Linux compile info:
https://holarse.de/wiki/hurrican

Would be great, if we could have this on our units :-)

Thx for reading / looking into it.

cu,
agerhard

**Comment by neilswann80** [comment]
I'd love to see this ported to OpenDingux, the Turrican series was always amazing.  Everything it's been ported to so far is different in architecture to our family of retro consoles (Ingenic JZ4770) however, so it wouldn't be an easy task.

Hopefully someone with the right skills (SDL and OpenGL at the very least) will be able to help.

In the meantime... I may try the "Super" and "Mega" console versions which I haven't played a lot of... Turrican was always an Amiga game to me!

**Comment by plrguez** [comment]
Hi,

I have built Hurrican for OD-beta. You can get it from here https://github.com/plrguez/hurrican-od/releases/tag/od-2022-10-16

Use the opk file `hurrican-od.opk` from the release page.

The other opk file is for the latest od firmware and it uses the new toolchain and probably in Adam you are using a previous OD firmware version.

**Reply by neilswann80** [reply]
Amazing work!

**Reply by neilswann80** [reply]
@plrguez do you think [Quake 3](https://github.com/podulator/ioq350) could be made to work on the new toolchain?

**Reply by plrguez** [reply]
I think that is possible. Seems to use EGL and needs rework for OD-Beta due to new drm/kms driver video, same as Hurrican. In fact, the code base is similar to Hurrican code int the EGL implementation.

I can give a try when I finish another thing that I am doing.

**Reply by SuperBadger77** [reply]
Do you need the new version of OD Beta on the Adam image to run the new Turrican?

**Reply by plrguez** [reply]
No, you can use this opk [`hurrican-od.opk`](https://github.com/plrguez/hurrican-od/releases/download/od-2022-10-16/hurrican-od.opk)

The other opk [` hurrican-od-perf.opk `](https://github.com/plrguez/hurrican-od/releases/download/od-2022-10-16/hurrican-od-perf.opk) is intended for use with the firmware published here https://github.com/OpenDingux/buildroot/actions/runs/3101426374

About this firmware version, and simplifying a lot, the performance boost need two things: the runtime (binutils in firmware) and that the programs compiled with the new toolchain. You can se the details about the performance boost in the [pull request](https://github.com/OpenDingux/buildroot/pull/81)

**Reply by neilswann80** [reply]
I tested both Hurrican opk files @plrguez uploaded and both work on Adam.  Not only do they work, but they play beautifully.

**Reply by plrguez** [reply]
@neilswann80 can you test this [Quake3 builds](https://github.com/plrguez/ioq3-od/releases/tag/20221022-od-v0.0.1)? 

Like Hurricane, there are two opks, corresponding to the new toolchain and the old one.

EDIT: I have updated the opk's to fix the display on RG280 devices

**Reply by neilswann80** [reply]
@plrguez Nice work!  `quake3-od_perf.opk` works on both my RG350M and my RG280v.  The RG350M everything looks pretty much perfect, but the scaling is off for the menus on the RG280v, only seeing the left half.  I managed by luck to start a game and the game itself is mostly scaled correctly (Seeing full image unlike the menus) but the bottom part of your weapon hasn't rendered and is see-through.

Amazing to have this running on the new firmware, it's loads of fun with the dual-sticks on the 350M.  😄

**Reply by plrguez** [reply]
Thank you! 

Try to redownload the `quake3-od_perf.opk` again, I have updated the opks in the same release [page](https://github.com/plrguez/ioq3-od/releases/tag/20221022-od-v0.0.1). It should solve the menu scaling issues on RG280. But the weapon rendering issue is still present.

I haven't go deep into how the game data is rendered, I just made the changes to include the DRM/KMS to display something on the screen. But for now I won't go beyond with this port, and I think it would probably be better to first invest in rebase it to the latest version of [ioquake3](https://github.com/ioquake/ioq3) and try to build for the OpenGL2 backend.

**Reply by neilswann80** [reply]
I think I downloaded and tested the opks afters you made the most recent edit/commit, but I've redownloaded and tested.  The menu screens are still only displaying the left half on the 280v.

**Reply by plrguez** [reply]
It's strange. I have just downloaded and installed both opk and it [display well](https://drive.google.com/file/d/1TWhuYnusuyaSaGawmPk_g7ze-m8Eqoqh/view?usp=sharing) on my RG280V.

Try to rename or delete the file `$HOME/.q3a/baseq3/q3config.cfg` and launch it again.

**Reply by neilswann80** [reply]
Renamed `q3config.cfg` and tried again, same issue.  Checked and a brand new `q3config.cfg` had been created.

**Reply by plrguez** [reply]
Try this new [build](https://drive.google.com/file/d/1Dx-wn60lBMF5SIJtbu_mh6y9E6U5gwCx/view?usp=sharing) to see if it solves the problem.

**Reply by neilswann80** [reply]
That didn't seem to fix it either, deleted the cfg file again also.  Don't worry too much though, its far better on the RG350M with dual-sticks anyhow!

**Reply by plrguez** [reply]
This is really puzzling. It can help to see the app log. Can you activate it, launch the app, deactivate it and show me the log? 

To activate the log don't know how to do from SimpleMenu, but in GMenu2x is using Start button from GMenu2x desktop and enable `Output logs`. The log file is `$HOME/.gmenu2x/gmenu2x.log`

And what version of od-beta are you using? I will try to install the same version to see if it is the problem.

**Reply by Shatur** [reply]
Same issue with Quake 3. I even tried to compiling it myself with your latest commit with the fix for RG280V, but looks like the item not appears in "Video mode". I even tried removing all other modes (from the place you added RG280V resolution) and recompiled it - no effect, I still can scroll all 12 items. So strange.

**Reply by Shatur** [reply]
I figured it out!
It works correctly with `ZTM’s Flexible HUD mod`. You probably have this mod installed.

**Reply by plrguez** [reply]
Sorry for late reply. You are right, there is a mod in my game data that corrects the menus for RG280V. So my fault for not testing it with the original game data :(, I have taken game data from an old sdcard used with stock firmware. 

I have tested now with the original game data and I got the problem. 

I have tried the patch **Keep Menu Aspect Ratio** from [here](https://www.unix-ag.uni-kl.de/~auerswal/quake3/index.html) but it doesn't solve the problem. So I'll keep investigating.

**Reply by plrguez** [reply]
I believe that I have a solution for this problem without the need of mods. 

Previously mentioned patch made me to look into how UI code is used and why the patch have no effect. By default UI is used via compiled QVM files contained in pk3 game files (compiled), and this code is used via shared libraries (native). To use the native mode ioquake must be launched with some parameters (or change the config file). See [the ioquake3 player's guide](https://ioquake3.org/help/players-guide/#sharedlibraries)

So I have made some code fixes and made a [new release](https://github.com/plrguez/ioq3-od/releases/tag/20221207-od-v0.0.2). You must use the opk file [quake3-native.opk](https://github.com/plrguez/ioq3-od/releases/download/20221207-od-v0.0.2/quake3-native.opk). 

You must rename/delete previous config file `/media/data/local/home/.q3a/baseq3/q3config.cfg` to avoid conflicts with new config in this release.

**Reply by Shatur** [reply]
Thank you, works like a charm.
